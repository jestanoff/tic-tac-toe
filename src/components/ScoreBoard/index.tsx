import * as React from 'react';
import * as PropTypes from 'prop-types';
import Score from 'components/Score';
import { SYMBOLS, PLAYER_X, PLAYER_O, M_DASH } from '../../constants';
import styles from './ScoresSection.css';

const ScoreBoard = ({ playerTurn, outcome }) => {
  const [state, setState] = React.useState({
    [PLAYER_X]: +localStorage.playerX || M_DASH,
    [PLAYER_O]: +localStorage.playerO || M_DASH,
  });

  React.useEffect(() => {
    const incrementScore = (score) => (score === M_DASH ? 1 : +score + 1);

    if (outcome === PLAYER_X) {
      localStorage.playerX = (+localStorage.playerX || 0) + 1;
      setState((prevState) => ({ ...prevState, [PLAYER_X]: incrementScore(prevState[PLAYER_X]) }));
    } else if (outcome === PLAYER_O) {
      localStorage.playerO = (+localStorage.playerO || 0) + 1;
      setState((prevState) => ({ ...prevState, [PLAYER_O]: incrementScore(prevState[PLAYER_O]) }));
    }
  }, [outcome]);

  const isScoreActive = React.useCallback(
    (player) => {
      if (outcome === 0) return false;
      return outcome > 0 ? outcome === player : playerTurn === player;
    },
    [outcome, playerTurn],
  );

  return (
    <section className={styles.container}>
      <Score score={state[PLAYER_X]} isActive={isScoreActive(PLAYER_X)} symbol={SYMBOLS[PLAYER_X]} />
      <Score score={state[PLAYER_O]} isActive={isScoreActive(PLAYER_O)} symbol={SYMBOLS[PLAYER_O]} />
    </section>
  );
};

ScoreBoard.propTypes = {
  outcome: PropTypes.number,
  playerTurn: PropTypes.number.isRequired,
};

ScoreBoard.defaultProps = {
  outcome: -1,
};

export default React.memo(ScoreBoard);
