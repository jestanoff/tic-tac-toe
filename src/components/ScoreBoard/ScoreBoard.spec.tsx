import * as React from 'react';
import { render, screen } from '@testing-library/react';
import ScoreBoard from './index';

describe('<ScoreBoard />', () => {
  afterEach(() => {
    localStorage.clear();
  });

  test('should render the score board with empty score for each player', () => {
    const { asFragment } = render(<ScoreBoard playerTurn={1} />);

    expect(screen.getByTitle('Player cross has 0 points')).toBeInTheDocument();
    expect(screen.getByTitle('Player circle has 0 points')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should increment the score by 1 when circle player wins', () => {
    render(<ScoreBoard playerTurn={1} outcome={1} />);

    expect(screen.getByTitle('Player cross has 1 points')).toBeInTheDocument();
    expect(screen.getByTitle('Player circle has 0 points')).toBeInTheDocument();
  });

  test('should increment the score by 1 when circle player wins', () => {
    render(<ScoreBoard playerTurn={2} outcome={2} />);

    expect(screen.getByTitle('Player cross has 0 points')).toBeInTheDocument();
    expect(screen.getByTitle('Player circle has 1 points')).toBeInTheDocument();
  });

  test('should get the score from localStorage', () => {
    localStorage.playerX = 10;
    localStorage.playerO = 99;
    render(<ScoreBoard playerTurn={1} />);

    expect(screen.getByTitle('Player cross has 10 points')).toBeInTheDocument();
    expect(screen.getByTitle('Player circle has 99 points')).toBeInTheDocument();
  });

  test('should persist the score into localStorage', () => {
    localStorage.playerX = 5;
    localStorage.playerO = 5;
    render(<ScoreBoard playerTurn={2} outcome={2} />);

    expect(screen.getByTitle('Player cross has 5 points')).toBeInTheDocument();
    expect(screen.getByTitle('Player circle has 6 points')).toBeInTheDocument();
    expect(localStorage.playerX).toBe('5');
    expect(localStorage.playerO).toBe('6');
  });

  test('should show the player that has to play their turn as active', () => {
    render(<ScoreBoard playerTurn={2} />);

    expect(screen.getByTestId('score-for-player-circle')).toHaveClass('active');
    expect(screen.getByTestId('score-for-player-cross')).toHaveClass('container');
  });
});
