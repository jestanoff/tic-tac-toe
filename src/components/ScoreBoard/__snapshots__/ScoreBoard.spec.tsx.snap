// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScoreBoard /> should render the score board with empty score for each player 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="active"
      data-testid="score-for-player-cross"
    >
      <span
        class="mark"
      >
        <div
          class="container"
        >
          <svg
            class="svg"
            data-testid="cross-mark"
            preserveAspectRatio="xMinYMin meet"
            version="1.1"
            viewBox="0 0 500 500"
          >
            <line
              stroke="#444444"
              stroke-width="50"
              x1="130"
              x2="130"
              y1="130"
              y2="130"
            >
              <animate
                attributeName="x2"
                attributeType="XML"
                dur="1ms"
                fill="freeze"
                from="130"
                id="first"
                repeatCount="1"
                to="370"
              />
              <animate
                attributeName="y2"
                attributeType="XML"
                dur="1ms"
                fill="freeze"
                from="130"
                repeatCount="1"
                to="370"
              />
            </line>
            <line
              opacity="0"
              stroke="#444444"
              stroke-width="50"
              x1="370"
              x2="370"
              y1="130"
              y2="130"
            >
              <animate
                attributeName="opacity"
                attributeType="XML"
                begin="0s"
                dur="1ms"
                fill="freeze"
                from="0"
                repeatCount="1"
                to="1"
              />
              <animate
                attributeName="x2"
                attributeType="XML"
                begin="0s"
                dur="1ms"
                fill="freeze"
                from="370"
                repeatCount="1"
                to="130"
              />
              <animate
                attributeName="y2"
                attributeType="XML"
                begin="0s"
                dur="1ms"
                fill="freeze"
                from="130"
                repeatCount="1"
                to="370"
              />
            </line>
          </svg>
        </div>
      </span>
      <span
        aria-live="polite"
        class="score"
        title="Player cross has 0 points"
      >
        —
      </span>
    </div>
    <div
      class="container"
      data-testid="score-for-player-circle"
    >
      <span
        class="mark"
      >
        <div
          class="container"
        >
          <svg
            class="svg"
            data-testid="circle-mark"
            preserveAspectRatio="xMinYMin meet"
            version="1.1"
            viewBox="0 0 500 500"
          >
            <circle
              cx="250"
              cy="250"
              fill="transparent"
              r="125"
              stroke="#444444"
              stroke-dasharray="800"
              stroke-dashoffset="-800"
              stroke-width="45"
            >
              <animate
                attributeName="stroke-dashoffset"
                attributeType="XML"
                dur="1ms"
                fill="freeze"
                from="-800"
                repeatCount="1"
                to="0"
              />
            </circle>
          </svg>
        </div>
      </span>
      <span
        aria-live="polite"
        class="score"
        title="Player circle has 0 points"
      >
        —
      </span>
    </div>
  </section>
</DocumentFragment>
`;
